### PR TITLE
[refactor] configure scoring parameters

### PR DESCRIPTION
**Pull request by akichidis**
## Description 

In testnet we have observed the phenomenon where the consensus scores get distributed to the full range of values between 0..300 which eventually makes the MED approach with the current configuration not flag any authority as low scoring , which is somewhat expected. However, we have good reasons to still exclude validators with low scoring (especially the cases < 10) . 

To further tune the parameters we pulled the reported scores of testnet via Grafana (we could equally get them via a validator's db, but that would require more effort) and introduced a test mainly for automation purposes to parse the scores (via a CSV file) , run the algorithm and study the output according to the newly tuned parameters. That allow us to play a little bit with the parameters and based on actual prod data get an idea of how aggressive the pruning is.

The PR is doing the following:
* tuning the scoring parameters to new values to better detected the spread values.
* introducing a slightly different implementation that doesn't disable the scoring mechanism when more than `f` low scoring authorities are detected. Now up to `f` authorities will flagged as low scoring, but no more.
* introducing a protocol config feature flag to enable the new low scoring detection mechanism
* introducing as protocol config the low scoring detection mechanism constants 

The output for the CSV input attached on this PR and the proposed parameters is the following:
```
"ANodes"                      36	28	--	--	--	42	--	--	--	--	23	29	--	--	--	53	--	--	54	76	37	--	--	29	42	40	63	--	39	--	--	--	--	
"Ankr"                        --	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	31	--	--	--	--	--	--	--	--	--	--	--	--	
"B-Harvest"                   --	--	49	--	30	--	--	--	30	56	--	--	--	--	--	37	--	--	--	35	--	--	--	--	56	--	--	35	--	--	41	--	--	
"BACKBONE"                    00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	00	
"BLRD"                        10	--	48	--	44	18	35	17	33	22	--	48	17	31	--	--	--	63	47	29	--	11	--	--	29	--	55	31	17	30	--	37	--	
"BiXinKelePool"               36	45	45	26	28	--	--	--	30	50	--	--	24	31	--	--	44	33	46	34	--	--	--	23	58	44	56	36	33	47	37	--	07	
"BlockEden.xyz"               --	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	52	--	--	--	--	--	--	--	--	--	
"BlockVision"                 45	26	--	22	--	36	--	--	58	--	--	--	--	41	--	--	--	17	--	43	--	--	--	--	--	--	--	--	--	--	--	--	--	
"Blockdaemon"                 41	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	37	--	--	--	--	--	--	--	40	44	--	--	--	--	--	--	
"Brightlystake"               41	--	28	--	--	44	--	--	47	34	--	31	14	17	--	38	--	30	33	62	--	--	18	--	42	41	58	--	13	46	21	17	38	
"BwareLabs"                   --	53	--	10	--	39	--	--	41	25	--	35	--	--	--	--	--	39	67	36	33	--	--	53	59	--	45	14	--	--	--	--	--	
"Chainbase"                   --	30	23	--	28	32	30	--	21	24	23	19	18	25	21	11	19	51	31	43	44	26	--	35	19	30	40	--	--	31	--	20	--	
"ComingChat"                  08	--	--	02	--	54	17	--	--	--	--	40	--	--	--	--	--	04	--	--	14	--	--	31	43	38	23	41	--	--	22	--	01	
"DSRV"                        37	--	40	--	--	55	21	23	41	37	--	26	--	31	--	36	--	35	34	46	22	--	--	--	--	--	--	--	--	--	--	--	32	
"Flipside"                    --	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	02	--	--	--	--	--	--	--	--	--	--	
"Forbole"                     22	29	38	--	--	26	--	--	33	--	--	--	--	22	--	40	--	46	--	37	--	--	--	--	54	--	58	--	--	42	--	--	--	
"Gaucho"                      --	49	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	37	--	--	--	--	52	--	--	--	--	--	--	--	40	
"H2O Nodes"                   --	--	--	--	--	--	--	--	--	46	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	
"HashQuark"                   27	35	48	17	50	23	--	--	38	56	--	--	--	26	21	51	--	--	57	50	--	--	20	--	43	--	40	--	--	--	--	--	--	
"Imperator"                   --	--	--	--	--	50	--	--	34	--	--	--	--	--	--	--	--	59	62	52	--	--	--	32	25	39	--	--	29	20	50	--	17	
"KingSuper"                   36	29	35	--	37	26	--	--	--	44	21	32	--	36	--	--	--	37	33	49	48	--	29	25	57	--	62	33	--	--	--	--	40	
"Moonlet"                     --	--	--	--	--	58	--	--	19	--	--	06	--	--	--	--	--	18	42	--	--	--	--	--	33	--	56	--	--	14	--	--	06	
"MoveFuns DAO"                --	--	--	28	50	45	--	--	41	44	24	43	23	34	--	--	22	31	66	24	--	--	--	--	23	--	63	--	25	44	33	17	--	
"Neuler"                      --	49	51	--	--	--	--	--	--	--	--	--	24	34	--	--	--	47	42	50	--	--	--	38	23	--	--	--	--	--	--	--	46	
"NodeReal"                    --	51	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	
"OKXEarn"                     41	59	35	22	--	10	31	--	26	44	--	--	--	22	16	34	42	48	29	34	39	25	--	32	35	14	28	15	24	28	--	21	28	
"ONBUFF"                      --	42	50	--	--	39	22	--	--	--	--	40	--	--	--	38	--	--	53	--	--	--	--	--	--	--	32	--	--	--	--	--	49	
"Overclock"                   --	51	39	--	--	--	--	--	--	--	15	--	21	--	--	--	--	--	--	--	--	--	--	51	40	--	--	--	31	--	--	--	--	
"Parasol"                     --	--	31	--	--	54	--	--	36	32	--	--	--	39	--	18	37	21	20	26	40	--	--	44	45	--	33	--	--	37	--	--	29	
"Qredo"                       --	46	--	--	--	59	--	11	43	40	--	12	--	37	--	33	03	27	--	--	44	--	--	35	00	00	00	--	--	--	14	--	--	
"SenseiNode"                  14	12	06	10	13	08	11	07	08	16	09	15	22	11	21	10	14	08	10	10	06	06	05	08	04	08	10	09	20	16	21	16	15	
"SpartanLabs"                 22	44	--	--	--	--	39	--	31	33	--	--	--	46	--	32	--	20	48	36	45	--	--	--	--	--	38	34	22	--	--	--	34	
"Stakin"                      --	54	--	--	--	54	--	--	23	--	--	--	--	--	--	53	37	28	46	--	--	--	--	--	33	37	--	--	--	18	--	35	--	
"XPRV"                        45	61	--	--	--	49	--	--	--	48	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	--	56	--	--	--	--	--	--	
"gumi"                        --	32	--	--	41	--	--	--	50	--	29	--	--	--	--	32	43	49	34	--	19	20	--	50	--	35	53	42	--	--	--	--	--	
"n1stake"                     00	00	00	00	00	00	00	00	00	43	--	28	00	00	21	--	09	57	--	--	--	--	--	--	--	--	--	--	--	--	09	--	--	
"staked"                      45	50	--	--	28	37	--	--	27	61	31	--	--	--	--	11	18	22	59	23	--	--	33	33	63	--	33	30	21	30	45	--	38	
---------------
Total                         18	23	17	10	12	24	10	06	23	20	09	15	10	18	06	17	12	25	22	22	14	06	07	17	24	13	23	12	12	14	11	08	16
```

where each line represents a host and it's corresponding score value . Each column represents the score values per host for each run of the algorithm for the given scores. It has to be mentioned that the provided CSV file contains on its first row the host names, and on every other line the validators scores snapshoted on the corresponding time instance.

Basically the above output can help us observe how many validators we prune on each "run" and for each host monitor their scores on the evolution of time . When no score value is present `--` , that means that the node has not been picked for pruning, thus we don't render the score.

## Test Plan 

CSV file of scores for a time window of ~1.5 hours as those are exported [by the metric](https://metrics.sui.io/d/FaNbUhene/sui-validators-narwhal-public-testnet-new?orgId=1&from=1681339104273&to=1681342704274&viewPanel=233)  

File [authority_scores.csv](https://github.com/MystenLabs/sui/files/11216856/authority_scores.csv)

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-core/src/scoring_decision.rs [hunk 1]
```diff
@@ -140,6 +177,115 @@ pub fn update_low_scoring_authorities(
...
+
+    // adjusted median absolute deviation
+    let mad = median(&abs_deviations) / mad_divisor;
+    let mut low_scoring = vec![];
+    for (i, (a, (score, _stake))) in scores_per_authority.iter().enumerate() {
+        let temp = deviations[i] / mad;
+
+        // We expect the methodology to include the zero scoring validators, but we explicitly
+        // include them to make sure, as we know for sure that those have no contribution.
+        if *score == 0 || temp < -cut_off_value {
```
Comment (akichidis) [inline crates/sui-core/src/scoring_decision.rs:245]: Always include in low scoring the zero score ones.

File: crates/sui-protocol-config/src/lib.rs [hunk 2]
```diff
@@ -135,6 +136,10 @@ struct FeatureFlags {
     // compatibility error.
     #[serde(skip_serializing_if = "is_false")]
     missing_type_is_compatibility_error: bool,
+    // If true, then the scoring decision mechanism will not get disabled when we do have more than
+    // f low scoring authorities, but it will simply flag as low scoring only up to f authorities.
+    #[serde(skip_serializing_if = "is_false")]
+    scoring_decision_with_no_disable: bool,
```
Comment (mystenmark) [inline crates/sui-protocol-config/src/lib.rs:142]: this flag name is a bit hard to understand. maybe `allow_bypass_scoring_decision`?

**Review by mystenmark — APPROVED** [review]
protocol changes seem good to me. my understanding is that this can't actually cause a fork anyway, right? it could just cause bad submission latency if different validators had different policies?

**Comment by akichidis** [comment]
> protocol changes seem good to me. my understanding is that this can't actually cause a fork anyway, right? it could just cause bad submission latency if different validators had different policies?

Exactly! And this would just last for a few hours until everyone eventually updated, but thought of this as a good opportunity to do it more gracefully since we do have the protocol config mechanism.

File: crates/sui-protocol-config/src/lib.rs (hunk 2 above)
Comment (akichidis) [reply crates/sui-protocol-config/src/lib.rs:142]: Maybe `scoring_decision_with_validity_threshold_cut_off` ?

**Pull request merged**